src/main/angular/app/services/graph-processor.service.spec.ts

```diff
@@ -0,0 +1,236 @@
+import {GraphProcessorService} from './graph-processor.service';
+import 'jasmine';
+import {DateUtilitiesService} from './date-utilities.service';
+import {GraphProperties} from '../../model/types';
+import {FakeDataService} from './fake-services/fake-data.service';
+import {SimpleChanges, SimpleChange} from '@angular/core';
+import {Project} from '../../model/project';
+import {ProjectGraphData} from '../../model/project-graph-data';
+
+describe('GraphProcessorService', () => {
+  let service: GraphProcessorService;
+
+  beforeEach(() => {
+    service = new GraphProcessorService(new DateUtilitiesService());
+  });
+
+  describe('initProperties()', () => {
+    it('Initializes empty data', () => {
+      const properties: GraphProperties = service.initProperties();
+
+      expect(properties.columns).toEqual(['Time']);
+      expect(properties.graphData).toEqual([]);
+      expect(properties.type).toBe('LineChart');
+      expect(properties.options.series).toBeDefined();
+    });
+  });
+
+  describe('processChanges()', () => {
+    let fakeDataService: FakeDataService;
+    let properties: GraphProperties;
+    let changes: SimpleChanges;
+
+    beforeAll(() => {
+      fakeDataService = new FakeDataService();
+    });
+
+    beforeEach(() => {
+      properties = service.initProperties();
+      changes = {};
+    });
+
+    describe('Adding projects to graph', () => {
+      it('Adds a single project successfully', async () => {
+        const project: Project = (await fakeDataService.listProjects())[0];
+        const projectData: ProjectGraphData = await fakeDataService.getProjectGraphData(
+          project.projectId
+        );
+
+        // Going from no projects to a single one
+        changes.projects = new SimpleChange([], [project], true);
+        await service.processChanges(changes, properties, fakeDataService);
+
+        expect(properties.columns.length).toBe(4);
+        expect(properties.columns[1]).toBe(project.projectId);
+
+        if (projectData) {
+          const bindingTimes: number[] = Object.keys(
+            projectData.dateToNumberIAMBindings
+          ).map(key => +key);
+          const recommendationTimes: number[] = Object.keys(
+            projectData.dateToRecommendationTaken
+          ).map(key => +key);
+          properties.graphData.forEach((row, index) => {
+            let time = -1;
+            // Make sure time is correct
+            if (row[0] instanceof Date) {
+              time = row[0].getTime();
+              expect(time).toBe(bindingTimes[index]);
+            } else {
+              fail('First row is not a time!');
+            }
+
+            // Make sure number bindings is correct
+            expect(row[1]).toBe(projectData.dateToNumberIAMBindings[time]);
```

Comment:
In general, it's best to have 1 expect per test case, so that when future eng are adding to this, and they are only changing a small piece of functionality, they can see why the test is failing. 

To accomplish this, you can split this into it's own declare block.